owsi-core/owsi-core-components/owsi-core-component-jpa/src/main/java/fr/openwide/core/jpa/batch/executor/MultithreadedBatch.java

```diff
@@ -0,0 +1,132 @@
+
+import java.math.BigDecimal;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.TransactionStatus;
+import org.springframework.transaction.support.TransactionCallback;
+
+import com.google.common.collect.Lists;
+
+import fr.openwide.core.jpa.batch.processor.ThreadedProcessor;
+import fr.openwide.core.jpa.batch.util.ProcessorProgressLogger;
+import fr.openwide.core.spring.property.service.IPropertyService;
+
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+public class MultithreadedBatch extends AbstractBatch<MultithreadedBatch> {
+	
+	private static final Logger LOGGER = LoggerFactory.getLogger(MultithreadedBatch.class);
+	
+	private static final Logger PROGRESS_LOGGER = LoggerFactory.getLogger(ProcessorProgressLogger.class);
+	
+	@Autowired
+	private IPropertyService propertyService;
+	
+	private int threads = 2;
+	
+	private int timeoutInMinutes = 15;
+	
+	public MultithreadedBatch threads(int threads) {
+		this.threads = threads;
+		return this;
+	}
+	
+	public MultithreadedBatch timeoutInMinutes(int timeout) {
+		this.timeoutInMinutes = timeout;
+		return this;
+	}
+	
+	public void run(String context, final List<Long> entityIds, final BatchRunnable<Long> batchRunnable) {
+		Date startTime = new Date();
+		
+		LOGGER.info("Beginning batch for %1$s: %2$d objects", context, entityIds.size());
+		
+		LOGGER.info("    preExecute start");
+		
+		writeTransactionTemplate.execute(new TransactionCallback<Void>() {
+			@Override
+			public Void doInTransaction(TransactionStatus status) {
+				batchRunnable.preExecute(entityIds);
+				return null;
+			}
+		});
+		
+		LOGGER.info("    preExecute end");
+
+		LOGGER.info("    starting batch executions");
+		
+		List<List<Long>> entityIdsPartitions = Lists.partition(entityIds, batchSize);
+		List<Callable<Void>> callables = Lists.newArrayList();
+		for (final List<Long> entityPartition : entityIdsPartitions) {
+			Callable<Void> callable = new Callable<Void>() {
+				@Override
+				public Void call() throws Exception {
+					batchRunnable.executePartition(entityPartition);
+					return null;
+				}
+			};
+			callables.add(callable);
+		}
+		
+		createThreadedProcessor(batchSize, timeoutInMinutes).runWithTransaction(context, callables, writeTransactionTemplate, entityIds.size());
+		
+		LOGGER.info("    end of batch executions");
+
+		LOGGER.info("    postExecute start");
+		
+		writeTransactionTemplate.execute(new TransactionCallback<Void>() {
+			@Override
+			public Void doInTransaction(TransactionStatus status) {
+				batchRunnable.postExecute(entityIds);
```

Comment:
There is a possibility that the first batch would have been executed just fine, while the other would have failed.
In this case, I guess an exception is thrown and this code is never executed, but the changes from the first batch likely have been persisted. Which is probably a problem.
Maybe we could provide a way to execute something on error? I mean, we could provide a way to define this inside the runnable? Say, `void postError(List<Long> entityIds)`?